wit/types.wit

```diff
@@ -1,44 +1,77 @@
-interface messaging-types {
+interface types {
     /// A connection to a message-exchange service (e.g., buffer, broker, etc.).
     resource client {
         connect: static func(name: string) -> result<client, error>;
     }
     
-    /// TODO(danbugs): This should be eventually extracted as an underlying type for other wasi-cloud-core interfaces.
-    resource error {
-        trace: static func() -> string;    
+    /// Errors that can occur when using the messaging interface.
+    variant error {
+        /// The requested option is not authorized. This could be a topic it doesn't have
+        /// permission to subscribe to, or a permission it doesn't have to perform a specific
+        /// action. This error is mainly used when calling `update-config`.
+        unauthorized,
+        /// The request or operation timed out.
+        timeout,
+        /// An error occurred with the connection. Includes a message for additional context
+        connection(string),
+        /// A catch all for other types of errors
+        other(string),
     }
   
-    /// There are two types of channels:
-    /// - publish-subscribe channel, which is a broadcast channel, and
-    /// - point-to-point channel, which is a unicast channel.
-    ///
-    /// The interface doesn't highlight this difference in the type itself as that's uniquely a consumer issue.
-    type channel = string;
+    /// The topic of a message. This is also called subject, channel, or group in various messaging
+    /// systems. What this value means is up to the implementation.
+    type topic = string;
   
-    /// Configuration includes a required list of channels the guest is subscribing to, and an optional list of extensions key-value pairs 
-    /// (e.g., partitions/offsets to read from in Kafka/EventHubs, QoS etc.).
-    record guest-configuration {
-        channels: list<channel>,
-        extensions: option<list<tuple<string, string>>>
+    /// Configuration includes a required list of channels the guest is subscribing to
+    record config {
+        channels: list<topic>,
     }
   
-    /// Format specification for messages 
-    ///  - more info: https://github.com/clemensv/spec/blob/registry-extensions/registry/spec.md#message-formats
-    ///  - message metadata can further decorate w/ things like format version, and so on.
-    enum format-spec {
-        cloudevents,
-        http,
-        amqp,
-        mqtt,
-        kafka,
-        raw
-    }
-  
-    /// A message with a binary payload, a format specification, and decorative metadata.
-    record message {
-        data: list<u8>,
-        format: format-spec,
-        metadata: option<list<tuple<string, string>>>
+    /// A message with a binary payload and additional information
+    resource message {
+        constructor(topic: topic, data: list<u8>, content-type: option<string>, metadata: option<list<tuple<string, string>>>);
```

Comment:
What are the expectations around the Message ID? Such metadata is so critical that having one way to figure this out would be good.

Reply:
Is a message ID actually important to a guest here? Most of the time those are used in acks or other advanced operations. Each implementation can take the wit type `message` and convert it to/assign it an ID based on implementation details IMO.

Reply:
From the guest's perspective, I think it is critical for the `guest` to control the identity of the message, the `guest` is most likely the component aware of the domain, while the `host` is at the platform level.

What the `host` would do with it shouldn't be part of the spec. That is a different story. Still, `guest` could find a component that provides the guarantees required around the dedupe and whatnot.

It's worth saying that the ID doesn't mean unique; it means an identifier.

Reply:
For whatever is worth, "messaging.message_id" OpenTelemetry Semantic Conventions Trace exists. Just to point to the important of such information

Reply:
I think I'd prefer to hold off and add this once we have a concrete use case for it